File: Source/NSArray+Sync.m
#import "NSArray+Sync.h"

#import "NSManagedObject+HYPPropertyMapper.h"
#import "NSManagedObject+Sync.h"

@implementation NSArray (Sync)

- (NSArray *)preprocessForEntityNamed:(NSString *)entityName
                       usingPredicate:(NSPredicate *)predicate
                               parent:(NSManagedObject *)parent
                            dataStack:(DATAStack *)dataStack {
    NSMutableArray *filteredChanges = [NSMutableArray new];

    if ([predicate isKindOfClass:[NSComparisonPredicate class]]) {
        NSComparisonPredicate *castedPredicate = (NSComparisonPredicate *)predicate;
        NSExpression *rightExpression = castedPredicate.rightExpression;
        id rightValue = [rightExpression constantValue];
        BOOL rightValueCanBeCompared = (rightValue &&
                                        ([rightValue isKindOfClass:[NSDate class]] ||
                                         [rightValue isKindOfClass:[NSNumber class]] ||
                                         [rightValue isKindOfClass:[NSString class]]));
        if (rightValueCanBeCompared) {
            NSMutableArray *objectChanges = [NSMutableArray new];
            NSEntityDescription *entity = [NSEntityDescription entityForName:entityName
                                                      inManagedObjectContext:dataStack.disposableMainContext];
            for (NSDictionary *change in self) {
                NSManagedObject *object = [[NSManagedObject alloc] initWithEntity:entity insertIntoManagedObjectContext:dataStack.disposableMainContext];
                NSError *error = nil;
                [object hyp_fillWithDictionary:change];
                [object sync_processRelationshipsUsingDictionary:change
                                                       andParent:parent
                                                       dataStack:dataStack
                                                           error:&error];
                [objectChanges addObject:object];
            }

            NSArray *filteredArray = [objectChanges filteredArrayUsingPredicate:predicate];
            for (NSManagedObject *filteredObject in filteredArray) {
                [filteredChanges addObject:[filteredObject hyp_dictionaryUsingRelationshipType:HYPPropertyMapperRelationshipTypeArray]];
            }
        }
    }

    return [filteredChanges copy];
}

@end

File: Source/NSArray+Sync.swift
import Foundation
import DATAStack
import NSManagedObject_HYPPropertyMapper
import DATAFilter

public extension NSArray {
  /**
   Filters the items using the provided predicate, useful to exclude JSON objects from a JSON array by using a predicate.
   - parameter entityName: The name of the entity to be synced.
   - parameter predicate: The predicate used to filter out changes, if you want to exclude some items, you just need to provide this predicate.
   - parameter parent: The parent of the entity, optional since many entities are orphans.
   - parameter dataStack: The DATAStack instance.
   */
  func preprocessForEntityNamed(_ entityName: String, predicate: Predicate, parent: NSManagedObject?, dataStack: DATAStack, operations: DATAFilterOperation) -> [[String : AnyObject]] {
    var filteredChanges = [[String : AnyObject]]()
    let validClasses = [NSDate.self, NSNumber.self, NSString.self]
    if let predicate = predicate as? ComparisonPredicate, selfArray = self as? [[String : AnyObject]] where validClasses.contains({ $0 as! NSObject == predicate.rightExpression.self }) {
      var objectChanges = [NSManagedObject]()
      let context = dataStack.newDisposableMainContext()
      if let entity = NSEntityDescription.entity(forEntityName: entityName, in: context) {
        selfArray.forEach {
          let object = NSManagedObject(entity: entity, insertInto: context)
          object.sync_fillWithDictionary($0, parent: parent, dataStack: dataStack, operations: operations)
          objectChanges.append(object)
        }

        guard let filteredArray = (objectChanges as NSArray).filtered(using: predicate) as? [NSManagedObject] else { fatalError("Couldn't cast filteredArray as [NSManagedObject]: \(objectChanges), predicate: \(predicate)") }
        for filteredObject in filteredArray  {
          if let change = filteredObject.hyp_dictionary(using: .array) as? [String : AnyObject] {
            filteredChanges.append(change)
          }
        }
      }
    }

    return filteredChanges
  }
}
